backpack/core/derivatives/lstm.py

```diff
@@ -0,0 +1,351 @@
+
+    @staticmethod
+    def _forward_pass(
+        module: LSTM, mat: Tensor
+    ) -> Tuple[Tensor, Tensor, Tensor, Tensor]:
+        """This performs an additional forward pass and returns the hidden variables.
+
+        This is important because the PyTorch implementation does not grant access to
+        some of the hidden variables. Those are computed and returned.
+
+        See also forward pass in class docstring.
+
+        Args:
+            module: module
+            mat: matrix, used to extract device and shapes
+
+        Returns:
+            ifgo, c, c_tanh, h
+        """
+        T: int = mat.shape[1]
+        N: int = mat.shape[2]
+        H: int = module.hidden_size
+        H0: int = 0 * H
+        H1: int = 1 * H
+        H2: int = 2 * H
+        H3: int = 3 * H
+        H4: int = 4 * H
+        # forward pass and save i, f, g, o, c, c_tanh-> ifgo, c, c_tanh
+        ifgo: Tensor = zeros(T, N, 4 * H, device=mat.device, dtype=mat.dtype)
+        c: Tensor = zeros(T, N, H, device=mat.device, dtype=mat.dtype)
+        c_tanh: Tensor = zeros(T, N, H, device=mat.device, dtype=mat.dtype)
+        h: Tensor = zeros(T, N, H, device=mat.device, dtype=mat.dtype)
+        for t in range(T):
+            ifgo[t] = (
+                einsum("hi,ni->nh", module.weight_ih_l0, module.input0[t])
+                + module.bias_ih_l0
+                + module.bias_hh_l0
+            )
+            if t != 0:
+                ifgo[t] += einsum(
+                    "hg,ng->nh", module.weight_hh_l0, module.output[t - 1]
+                )
+            ifgo[t, :, H0:H1] = sigmoid(ifgo[t, :, H0:H1])
+            ifgo[t, :, H1:H2] = sigmoid(ifgo[t, :, H1:H2])
+            ifgo[t, :, H2:H3] = tanh(ifgo[t, :, H2:H3])
+            ifgo[t, :, H3:H4] = sigmoid(ifgo[t, :, H3:H4])
+            c[t] = ifgo[t, :, H0:H1] * ifgo[t, :, H2:H3]
+            if t != 0:
+                c[t] += ifgo[t, :, H1:H2] * c[t - 1]
+            c_tanh[t] = tanh(c[t])
+            h[t] = ifgo[t, :, H3:H4] * c_tanh[t]
+
+        return ifgo, c, c_tanh, h
+
+    @classmethod
+    def _ifgo_jac_t_mat_prod(cls, module: LSTM, mat: Tensor) -> Tensor:
+        V: int = mat.shape[0]
+        T: int = mat.shape[1]
+        N: int = mat.shape[2]
+        H: int = module.hidden_size
+        H0: int = 0 * H
+        H1: int = 1 * H
+        H2: int = 2 * H
+        H3: int = 3 * H
+        H4: int = 4 * H
+
+        ifgo, c, c_tanh, _ = cls._forward_pass(module, mat)
+
+        # backward pass
+        H_prod_t: Tensor = zeros(V, N, H, device=mat.device, dtype=mat.dtype)
+        C_prod_t: Tensor = zeros(V, N, H, device=mat.device, dtype=mat.dtype)
+        C_prod_old: Tensor = zeros(V, N, H, device=mat.device, dtype=mat.dtype)
+        IFGO_prod: Tensor = zeros(V, T, N, 4 * H, device=mat.device, dtype=mat.dtype)
+        for t in reversed(range(T)):
+            # jac_t_mat_prod until node h
+            H_prod_t[:] = mat[:, t]
```

Comment:
I assume that the `[:]` slice is necessary here as otherwise `H_prod_t` points to the same memory location as `mat`, and the following inplace summation will thus alter `mat`, but shouldn't.

To make this more explicit, what do you think about replacing the `H_prod_t = zeros(...)` line + `H_prod_t[:]` syntax by a more explicit `H_prod_t = mat[:, t].clone()`?

Reply:
A similar argument applies to `C_prod` and `C_prod_old`. Mentioning it here to keep the discussion in a single conversation.

Reply:
My intuition (based on what I learned from other programming languages):
`H_prod_t = mat[:, t].clone()` creates a copy of `mat` and then sets the pointer of `H_prod_t` to that copy. Finally, the space previously allocated for `H_prod_t` is freed.
`H_prod_t[:] = mat[:, t]` changes the content of the memory of `H_prod_t` to the values of `mat`.

Do you agree?
If no, I will try to read up whether my intuition is based on anything factual.

Reply:
Agree that your solution only allocates memory once, in contrast to my proposed change.